Templatetag remplacant une espace par une espace insécable

helmasaur opened the pull request:
Permet de replace un espace par un espace insécable devant ou après certains caractères.

Numéro du ticket concerné : close #5292

### Contrôle qualité

1. Tester le nouveau filtre avec la commande : `python manage.py test --settings zds.settings.test zds.utils.tests.tests_french_typography`
2. Mettre un contenu en Une avec le titre suivant : `Comment alerter les secours ?`
3. Voir si le problème de l'illustration ci-dessous est corrigé
4. Vérifier que cela n'engendre pas de nouvelles erreurs d'affichage

<img width="228" alt="53592687-1af6e780-3b97-11e9-951b-054ae0d9f67c" src="https://user-images.githubusercontent.com/873762/87880248-817cdf80-c9f0-11ea-9e86-78c0f5cf2c4f.png">

helmasaur committed:
✨ Ajout du templatetag remplaçant les espaces par des espaces insécables

helmasaur commented:
J'ai fait une tentative de _templatetag_ mais je ne sais pas si ce que j'ai fait est bien fait… Je ne sais pas non plus si j'ai choisi un bon nom de fichier.

helmasaur committed:
🎨 Utilisation d'espaces pour l'indentation

AmauryCarrade commented:
Un filtre me semble être la bonne solution pour ce genre de choses. Par contre, il serait bien d'avoir un test associé :) .

Aussi, je me demande s'il ne serait pas mieux d'utiliser une entité HTML (`&nbsp;`), mais il faut alors [marquer la sortie comme sûre](https://docs.djangoproject.com/en/3.0/ref/utils/#django.utils.safestring.mark_safe).

Enfin, c'est un filtre sur une chaîne de caractères, donc [il est recommandé d'utiliser en plus `@stringfilter`](https://docs.djangoproject.com/en/3.0/howto/custom-template-tags/#template-filters-that-expect-strings), afin que Django s'assure qu'une chaîne de caractère est toujours passée (convertissant automatiquement les autres types).

helmasaur committed:
🎨 Amélioration du code

- Remplacement de la dernière tabulation utilisée par des espaces
- ajout de `@stringfilter`
- utilisation des entités HTML à la place des caractères UTF-8

helmasaur committed:
✅ Ajout du test du filtre non_breaking_space

helmasaur committed:
🎨 Clarification du code et correction d'une coquille

helmasaur committed:
✏️ Correction d'une coquille

helmasaur committed:
💄 Application du filtre au titre des Unes

helmasaur commented:
Le problème est corrigé pour les Unes à la page d'accueil. Par contre, je ne sais pas comment corriger le même problème sur la page de mise en Une. C'est moins grave car seuls les membres du staff ont accès à cette page. Le problème est l'utilisation de JavaScript pour générer l'aperçu.

Si ce problème n'est pas un problème immédiat, la PR est prêt et on peut ouvrir un autre ticket pour le problème expliqué ci-dessus.

AmauryCarrade commented:
> Par contre, je ne sais pas comment corriger le même problème sur la page de mise en Une. C'est moins grave car seuls les membres du staff ont accès à cette page. Le problème est l'utilisation de JavaScript pour générer l'aperçu.

À part réimplémenter le filtre en JS, je ne vois pas comment ce serait possible (sauf solutions _overkill_ du genre faire le rendu côté serveur mais je pense que c'est une mauvaise idée, pas du tout pertinente juste pour une histoire d'espaces insécables).

AmauryCarrade commented on zds/utils/tests/test_non_breaking_space.py:
```diff
@@ -0,0 +1,38 @@
+        test_cases = ReplaceNonBreakingSpace.get_cases()
+        for case_name, case in test_cases.items():
+                with self.subTest(msg=case_name):
+                    self.assertEqual(case['expected_output'], non_breaking_space(case['input']))
```
Il faudrait aussi tester d'autres cas, par exemple : 
- un texte sans aucun élément à changer ne doit pas être modifié ;
- un texte avec plusieurs éléments identiques transformés, doit voir tous ces éléments transformés ;
- un texte avec plusieurs éléments différents transformés doit voir tous ces éléments transformés ;
- un texte vide ou `None` doit être transformé en un texte vide.

AmauryCarrade commented on zds/utils/templatetags/non_breaking_space.py:
```diff
@@ -0,0 +1,20 @@
+
+register = template.Library()
+
+@register.filter('non_breaking_space')
```
Inutile de spécifier un nom au filtre si c'est le même nom que la fonction.

AmauryCarrade commented on zds/utils/templatetags/non_breaking_space.py:
```diff
@@ -0,0 +1,20 @@
+
+@register.filter('non_breaking_space')
+@stringfilter
+def non_breaking_space(str):
```
Je me demande si c'est le meilleur nom ; par exemple on pourrait l'appeler `french_typography` (plus représentatif) ?

AmauryCarrade reviewed (changes requested):
En plus des problèmes remontés, Travis n'est pas contente car il y a des soucis de lint. Pas mal de soucis de mélanges de tabulations et d'espaces, et quelques autres soucis ; je te laisse regarder ci-dessous (ou [dans le log de Travis](https://travis-ci.com/github/zestedesavoir/zds-site/jobs/362548959#L2084-L2101)).

```
./zds/utils/tests/test_non_breaking_space.py:5:1: E302 expected 2 blank lines, found 1
./zds/utils/tests/test_non_breaking_space.py:12:17: E128 continuation line under-indented for visual indent
./zds/utils/tests/test_non_breaking_space.py:15:17: E128 continuation line under-indented for visual indent
./zds/utils/tests/test_non_breaking_space.py:18:17: E128 continuation line under-indented for visual indent
./zds/utils/tests/test_non_breaking_space.py:21:17: E128 continuation line under-indented for visual indent
./zds/utils/tests/test_non_breaking_space.py:25:17: E128 continuation line under-indented for visual indent
./zds/utils/tests/test_non_breaking_space.py:28:17: E128 continuation line under-indented for visual indent
./zds/utils/tests/test_non_breaking_space.py:31:17: E128 continuation line under-indented for visual indent
./zds/utils/tests/test_non_breaking_space.py:37:17: E117 over-indented
./zds/utils/tests/test_non_breaking_space.py:38:97: W292 no newline at end of file
./zds/utils/templatetags/non_breaking_space.py:7:1: E302 expected 2 blank lines, found 1
./zds/utils/templatetags/non_breaking_space.py:12:1: W191 indentation contains tabs
./zds/utils/templatetags/non_breaking_space.py:12:1: E101 indentation contains mixed spaces and tabs
./zds/utils/templatetags/non_breaking_space.py:12:3: E131 continuation line unaligned for hanging indent
./zds/utils/templatetags/non_breaking_space.py:13:1: W191 indentation contains tabs
./zds/utils/templatetags/non_breaking_space.py:13:1: E101 indentation contains mixed spaces and tabs
./zds/utils/templatetags/non_breaking_space.py:14:1: W191 indentation contains tabs
./zds/utils/templatetags/non_breaking_space.py:14:1: E101 indentation contains mixed space
```

helmasaur committed:
🎨 Amélioration du code

- indentation n'utilisant plus que des espaces
- suppression du nom du filtre
- ajout de la ligne vide finale

helmasaur committed:
🎨 Ajout de lignes vides

Pour faire plaisir à Travis

helmasaur committed:
🎨 Dernières corrections pour Travis

helmasaur commented:
J'ai fini les modifications destinées à faire plaisir à Travis. Je m'occuperai ensuite des modifications demandées.

helmasaur committed:
🎨 Ajout d'un saut de ligne pour Travis

helmasaur committed:
✅ Nouveau test et nom des tests en anglais

helmasaur committed:
🚚 Renommage en french_typography

helmasaur commented:
J'ai renommer le filtre et j'ai ajouté les tests qui m'ont été demandé. Il reste à savoir si je fais le même filtre en JavaScript dans cette PR ou dans une autre.

>  Par contre, je ne sais pas comment corriger le même problème sur la page de mise en Une. C'est moins grave car seuls les membres du staff ont accès à cette page. Le problème est l'utilisation de JavaScript pour générer l'aperçu.

AmauryCarrade commented on zds/utils/templatetags/french_typography.py:
```diff
@@ -0,0 +1,21 @@
+
+@register.filter()
+@stringfilter
+def french_typography(str):
```
De la documentation pour ce filtre serait la bienvenue. (Pourquoi tu colories n'importe comment, GitHub ?)

```suggestion
def french_typography(str):
    """
    Replaces spaces with non-breaking-spaces or narrow non-breaking-spaces
    before or after some symbols, according to French typography.
    
    This filter is naive and should not be used on Markdown content.
    """
```

AmauryCarrade reviewed (changes requested):
Ça manque de doc, mais sinon on est pas mal !

> Il reste à savoir si je fais le même filtre en JavaScript dans cette PR ou dans une autre.

Pour moi, vu que c'est uniquement pour le _staff_ et dans un cas rare, c'est loin d'être urgent. Si ça t'amuse de le faire (je comprendrais :p), vas-y et ce sera accepté avec plaisir (dans une autre PR, je pense) ; mais c'est pas un drame si ce n'est pas fait. :)

AmauryCarrade reviewed (approved):
:heavy_check_mark: QA OK (cette approbation ne retire pas la demande précédente, bien sûr)

helmasaur committed:
💡 Commentaire explicant le fonctionnement du filtre

Co-authored-by: Amaury Carrade <amaury@carrade.eu>

AmauryCarrade committed:
Merge branch 'dev' into 5292-templatetag-espace-insecable

helmasaur committed:
🎨 Suppression des espaces ajoutés par GitHub
